db/db.go

```diff
@@ -16,52 +25,225 @@ type Scannable interface {
+			"ClientCollateral":      &bigIntFieldDef{f: &deal.ClientDealProposal.Proposal.ClientCollateral},
+			"SelfPeerID":            &fieldDef{f: &deal.SelfPeerID},
+			"ClientPeerID":          &fieldDef{f: &deal.ClientPeerID},
+			"DealDataRoot":          &cidFieldDef{f: &deal.DealDataRoot},
+			"InboundFilePath":       &fieldDef{f: &deal.InboundFilePath},
+			"TransferType":          &fieldDef{f: &deal.TransferType},
+			"TransferParams":        &fieldDef{f: &deal.TransferParams},
+			"ChainDealID":           &fieldDef{f: &deal.ChainDealID},
+			"PublishCID":            &cidFieldDef{f: deal.PublishCID},
+			"SectorID":              &fieldDef{f: &deal.SectorID},
+			"Offset":                &fieldDef{f: &deal.Offset},
+			"Length":                &fieldDef{f: &deal.Length},
+			"Checkpoint":            &ckptFieldDef{f: &deal.Checkpoint},
+			"Error":                 &fieldDef{f: &deal.Err},
+		},
+	}
+}
+
+func (d *dealAccessor) scan(row Scannable) error {
+	// For each field
+	dest := []interface{}{}
+	for _, name := range dealFields {
+		// Get a pointer to the field that will receive the scanned value
+		fieldDef := d.def[name]
+		dest = append(dest, fieldDef.fieldPtr())
+	}
+
+	// Scan the row into each pointer
+	err := row.Scan(dest...)
+	if err != nil {
+		return xerrors.Errorf("scanning deal row: %w", err)
+	}
+
+	// For each field
+	for _, fieldDef := range d.def {
+		// Unmarshall the scanned value into deal object
+		err := fieldDef.unmarshall()
+		if err != nil {
+			return xerrors.Errorf("unmarshalling db field: %s", err)
+		}
+	}
+	return nil
+}
+
+func (d *dealAccessor) insert(ctx context.Context) error {
+	// For each field
+	values := []interface{}{}
+	placeholders := make([]string, 0, len(values))
+	for _, name := range dealFields {
+		// Add a placeholder "?"
+		fieldDef := d.def[name]
+		placeholders = append(placeholders, "?")
+
+		// Marshall the field into a value that can be stored in the database
+		v, err := fieldDef.marshall()
+		if err != nil {
+			return err
+		}
+		values = append(values, v)
+	}
+
+	// Execute the INSERT
+	qry := "INSERT INTO Deals (" + dealFieldsStr + ") "
+	qry += "VALUES (" + strings.Join(placeholders, ",") + ")"
+	_, err := d.db.ExecContext(ctx, qry, values...)
+	return err
+}
+
+func (d *dealAccessor) update(ctx context.Context) error {
+	// For each field
+	values := []interface{}{}
+	setNames := make([]string, 0, len(values))
+	for _, name := range dealFields {
+		// Skip the ID field
+		if name == "ID" {
+			continue
+		}
+
+		// Add "fieldName = ?"
+		fieldDef := d.def[name]
+		setNames = append(setNames, name+" = ?")
+
+		// Marshall the field into a value that can be stored in the database
+		v, err := fieldDef.marshall()
+		if err != nil {
+			return err
+		}
+		values = append(values, v)
+	}
+
+	// Execute the UPDATE
+	qry := "UPDATE Deals "
+	qry += "SET " + strings.Join(setNames, ", ")
+
+	qry += "WHERE ID = ?"
+	values = append(values, d.deal.DealUuid)
```

Comment:
I don't understand how we specify the dealUUID in the WHERE part of the clause here. Why do we append the `DealUuid` at the end here ?

Reply:
Because we need to replace the "?" with the dealUUID in "WHERE ID = ?".
If we don't have a WHERE clause we will override all rows in the DB with the same data.